src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs

```diff
@@ -167,43 +157,149 @@ internal int EndRead(IAsyncResult asyncResult)
+
+            return WriteAsyncInternal(writeAdapter, buffer, offset, count);
+        }
+
+        private Task WriteAsyncInternal<T>(T writeAdapter, byte[] buffer, int offset, int count)
+            where T : struct, ISslWriteAdapter
+        {
+            _sslState.CheckThrow(authSuccessCheck: true, shutdownCheck: true);
+
+            if (count == 0 && !SslStreamPal.CanEncryptEmptyMessage)
             {
-                throw new ArgumentNullException(nameof(asyncResult));
+                // If it's an empty message and the PAL doesn't support that, we're done.
+                return Task.CompletedTask;
             }
 
-            LazyAsyncResult lazyResult = asyncResult as LazyAsyncResult;
-            if (lazyResult == null)
+            if (Interlocked.Exchange(ref _nestedWrite, 1) == 1)
             {
-                throw new ArgumentException(SR.Format(SR.net_io_async_result, asyncResult.GetType().FullName), nameof(asyncResult));
+                throw new NotSupportedException(SR.Format(SR.net_io_invalidnestedcall, "WriteAsync", "write"));
             }
 
-            if (Interlocked.Exchange(ref _nestedWrite, 0) == 0)
+            if (count <= _sslState.MaxDataSize)
             {
-                throw new InvalidOperationException(SR.Format(SR.net_io_invalidendcall, "EndWrite"));
+                //Single write
+                Task t = WriteSingleChunk(writeAdapter, buffer, offset, count);
+                if (t.IsCompleted)
+                {
+                    _nestedWrite = 0;
+                    return t;
+                }
+                else
+                {
+                    return ExitWriteAsync(t);
+                }
+            }
+            else
+            {
+                return ExitWriteAsync(WriteAsyncChunked(writeAdapter, buffer, offset, count));
             }
 
-            // No "artificial" timeouts implemented so far, InnerStream controls timeout.
-            lazyResult.InternalWaitForCompletion();
-
-            if (lazyResult.Result is Exception e)
+            async Task ExitWriteAsync(Task task)
             {
-                if (e is IOException)
+                try
                 {
-                    ExceptionDispatchInfo.Throw(e);
+                    await task.ConfigureAwait(false);
                 }
+                catch (Exception e)
+                {
+                    _sslState.FinishWrite();
+
+                    if (e is IOException)
+                    {
+                        throw;
+                    }
 
-                throw new IOException(SR.net_io_write, e);
+                    throw new IOException(SR.net_io_write, e);
+                }
+                finally
+                {
+                    _nestedWrite = 0;
+                }
             }
         }
 
+        private Task WriteSingleChunk<T>(T writeAdapter, byte[] buffer, int offset, int count)
+            where T : struct, ISslWriteAdapter
```

Comment:
Nit: often "T" is used to represent data elements.  I think it'd be nice to be more descripive here, e.g. TWriteAdapter.  Same for any other places this pattern is used.

Reply:
Changed